packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js

```diff
@@ -1,621 +1,944 @@
-          minSponsorTokens: { rawValue: convertSynthetic("1") },
-          timerAddress: timer.address,
-          excessTokenBeneficiary: store.address,
-          financialProductLibraryAddress: ZERO_ADDRESS
-        };
-
-        // The ExpiringMultiPartyClient does not emit any info `level` events.  Therefore no need to test Winston outputs.
-        // DummyLogger will not print anything to console as only capture `info` level events.
-        const dummyLogger = winston.createLogger({
-          level: "info",
-          transports: [new winston.transports.Console()]
-        });
+// These unit tests are re-run against the array of contract types and versions below. unit tests can choose which version
+// they support using the `versionedIt` syntax. Additional versions can be added, once an UMA release has been done.
+const SUPPORTED_CONTRACT_VERSIONS = ["ExpiringMultiParty-1.2.2", "ExpiringMultiParty-latest", "Perpetual-latest"];
+
+const startTime = "15798990420";
+const zeroAddress = "0x0000000000000000000000000000000000000000";
+const unreachableDeadline = MAX_UINT_VAL;
+
+// Common contract objects.
+let collateralToken;
+let syntheticToken;
+let emp;
+let mockOracle;
+let identifierWhitelist;
+let identifier;
+let finder;
+let store;
+let timer;
+let collateralWhitelist;
+let constructorParams;
+let currentTestIterationVersion;
+
+// Js Objects, clients and helpers
+let client;
+let convertCollateral;
+let convertSynthetic;
+let convertPrice;
+let dummyLogger;
+
+// Perpetual
+let configStore;
+let optimisticOracle;
+let fundingRateIdentifier;
+
+// Helper functions
+// TODO figure out the best pattern to refactor these into a library to make them re-usable in other tests.
+const updateAndVerify = async (client, expectedSponsors, expectedPositions) => {
+  await client.update();
+  assert.deepStrictEqual(client.getAllSponsors().sort(), expectedSponsors.sort());
+  assert.deepStrictEqual(client.getAllPositions().sort(), expectedPositions.sort());
+};
+
+// Set the funding rate and advances time by 10k seconds.
+const _setFundingRateAndAdvanceTime = async fundingRate => {
+  const currentTime = (await emp.getCurrentTime()).toNumber();
+  await emp.proposeFundingRate({ rawValue: fundingRate }, currentTime);
+  await emp.setCurrentTime(currentTime + 10000);
+};
+
+// If the current version being executed is part of the `supportedVersions` array then return `it` to run the test.
+// Else, do nothing. Can be used exactly in place of a normal `it` to parameterize contract types and versions supported
+// for a given test.eg: versionedIt(["Perpetual-latest"])("test name", async function () { assert.isTrue(true) })
+// Note that a second param can be provided to make the test an `it.only` thereby ONLY running that single test, on
+// the provided version. This is very useful for debugging and writing single unit tests without having ro run all tests.
+const versionedIt = function(supportedVersions, shouldBeItOnly = false) {
```

Comment:
this function is meant to replace the mocha `it` through the tests by providing a means to define a set of "contract versions" that is supported.

Reply:
in the next PR I will pull this to a common helper util.

Reply:
note that the second param here is optional. if included it will make the test exclusive (`it.only`). This could be dangerous syntax as it will exclude other tests within a test file but won't be caught by our linter. Am curious what others think about this.

Reply:
Do you use the `shouldBeItOnly` on any production tests? or is this only for debugging

Reply:
only for debugging. it acts as `it.only`

Reply:
`versionedIt` and `runTestForContractVersion` seem like they'd be useful for other clients. do you plan on pulling them into a shared file eventually?

Reply:
Also, does this mean that the linter will no longer catch any `it.only()` tests leftover from testing?

Reply:
yup. will pull them into a shared util in the next PR.

and you right; the linter will no longer catch `it.only`s. if you were to put `versionedIt(["any"], true)` it be equivalent to `it.only` but the linter will miss this. I will look into a custom linter rule to catch this.